src/test/server_test.cpp

```diff
@@ -0,0 +1,121 @@
+using testing::Return;
+using testing::ReturnRef;
+using testing::UnorderedElementsAre;
+
+using grpc::Channel;
+using grpc::ClientContext;
+
+struct Configuration {
+    std::string address = "localhost";
+    std::string port = "9178";
+};
+
+class ServingClient {
+    std::unique_ptr<inference::GRPCInferenceService::Stub> stub_;
+
+public:
+    ServingClient(std::shared_ptr<Channel> channel) :
+        stub_(inference::GRPCInferenceService::NewStub(channel)) {
+    }
+
+    // Pre-processing function for synthetic data.
+    // gRPC request proto is generated with synthetic data with shape/precision matching endpoint metadata.
+    void verifyLive(grpc::StatusCode expectedStatus = grpc::StatusCode::OK, bool alive = true) {
+        ClientContext context;
+        ::inference::ServerLiveRequest request;
+        ::inference::ServerLiveResponse response;
+
+        auto status = stub_->ServerLive(&context, request, &response);
+        ASSERT_EQ(status.error_code(), expectedStatus);
+        EXPECT_EQ(response.live(), alive);
+    }
+
+private:
+    ::inference::ServerLiveRequest request;
+};
+
+void requestServerAlive(const char* grpcPort, grpc::StatusCode status = grpc::StatusCode::OK, bool expectedStatus = true) {
+    grpc::ChannelArguments args;
+    std::string address = std::string("localhost") + ":" + grpcPort;
+    ServingClient client(grpc::CreateCustomChannel(address, grpc::InsecureChannelCredentials(), args));
+    client.verifyLive(status, expectedStatus);
+}
+
+TEST(Server, ServerNotAliveBeforeStart) {
+    // here we should fail to connect before starting server
+    requestServerAlive("9178", grpc::StatusCode::UNAVAILABLE, false);
+}
+
+TEST(Server, ServerAliveBeforeLoadingModels) {
+    // purpose of this test is to ensure that the server responds with alive=true before loading any models.
+    // this is to make sure that eg. k8s won't restart container until all models are loaded because of not being alivea
+    std::string port = "9000";
+    const char* testPort = port.c_str();
+    std::mt19937_64 eng{std::random_device{}()};
+    std::uniform_int_distribution<> dist{0, 9};
+    std::this_thread::sleep_for(std::chrono::milliseconds{dist(eng)});
+    for (auto j : {1, 2, 3}) {
+        char* digitToRandomize = (char*)testPort + j;
+        *digitToRandomize += dist(eng);
+    }
+
+    char* argv[] = {
+        (char*)"OpenVINO Model Server",
+        (char*)"--model_name",
+        (char*)"dummy",
+        (char*)"--model_path",
+        (char*)"/ovms/src/test/dummy",
+        (char*)"--port",
+        (char*)port.c_str(),
+        nullptr};
+    requestServerAlive(argv[6], grpc::StatusCode::UNAVAILABLE, false);
+    ovms::Server& server = ovms::Server::instance();
+    std::thread t([&argv, &server]() {
+        server.start(7, argv);
+    });
+    auto start = std::chrono::high_resolution_clock::now();
+    while ((ovms::Server::instance().getModuleState("GRPCServerModule") != ovms::ModuleState::INITIALIZED) &&
```

Comment:
This test will hang if model will fail to load

Reply:
No -  there is failsafe - we wait max 1s.